core/dom/element.js

```diff
@@ -419,28 +419,127 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
+						forEach = CKEDITOR.tools.array.forEach,
+						lastElement,
+						scrollables;
+
+					if ( isScrollLockFallbackNeeded && focusOptions && focusOptions.preventScroll ) {
+						scrollables = storeScrollPosition( element );
+					} else if ( CKEDITOR.env.chrome || CKEDITOR.env.ie ) {
+						lastElement = !editable || editable.isInline() ? element : editable.getParent();
+						scrollables = storeScrollPosition( element, lastElement );
+					}
 
-			return function( defer ) {
-				if ( defer )
-					CKEDITOR.tools.setTimeout( exec, 100, this );
-				else
-					exec.call( this );
-			};
-		} )(),
+					if ( scrollables && CKEDITOR.env.ie ) {
+						var scrolledElements = [ CKEDITOR.document.getWindow(), element ];
+
+						if ( scrolledElements[ 0 ] !== element.getWindow() ) {
+							scrolledElements.push( element.getWindow() );
+						}
+
+						forEach( scrolledElements, function( element ) {
+							var listener = element.once( 'scroll', function() {
+								if ( CKEDITOR.env.version < 9 ) {
+									setTimeout( function() {
+										restoreScrollAndCleanScrollables( scrollables );
+									} );
+								} else {
+									restoreScrollAndCleanScrollables( scrollables );
+								}
+							} );
+
+							// Clean listener to not break user scroll.
+							setTimeout( function() {
+								listener.removeListener();
+							}, 50 );
+						} );
+					}
+
+					// [IE] Use instead "setActive" method to focus the editable if it belongs to the host page document,
+					// to avoid bringing an unexpected scroll.
+					if ( CKEDITOR.env.ie && !( CKEDITOR.env.edge && CKEDITOR.env.version > 14 ) && this.getDocument().equals( CKEDITOR.document ) ) {
+						this.$.setActive();
+					} else {
+						this.$.focus( focusOptions );
+					}
+
+					if ( scrollables && !CKEDITOR.env.ie ) {
+						restoreScrollPosition( scrollables );
+					}
+				} catch ( e ) {
+					// IE throws unspecified error when focusing editable after closing dialog opened on nested editable.
+					if ( !CKEDITOR.env.ie ) {
+						throw e;
+					}
+				}
+
+				function storeScrollPosition( element, lastElement ) {
+					var scrollables = [];
+					while ( element ) {
+						var isElementScrollable = element.$.scrollHeight > element.$.clientHeight || element.getName() == 'body';
+
+						if ( isElementScrollable ) {
+							scrollables.push( {
+								element: element,
+								scrollTop: element.$.scrollTop
+							} );
+						}
+
+						if ( element.equals( lastElement ) ) {
+							break;
+						}
+
+						element = element.getParent() || element.getWindow().getFrame();
+					}
+
+					return scrollables;
+				}
+
+				function restoreScrollPosition( scrollables ) {
+					forEach( scrollables, function( item ) {
+						item.element.$.scrollTop = item.scrollTop;
+					} );
+				}
+
+				function restoreScrollAndCleanScrollables( scrollables ) {
+					restoreScrollPosition( scrollables );
+					scrollables = [];
```

Comment:
This line does not clean `scrollables` variable from outer scope, it's just change object referenced by _local_ `scrollables` variable. To really clean `scrollables`, `scrollables.length = 0` could be used.